src/shogun/structure/GEMPLP.h

```diff
@@ -0,0 +1,160 @@
+#define __GEMPLP_H__
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/structure/FactorGraph.h>
+#include <shogun/structure/Factor.h>
+#include <shogun/structure/MAPInference.h>
+#include <shogun/lib/SGNDArray.h>
+
+#include <vector>
+
+using namespace std;
+
+namespace shogun
+{
+#define IGNORE_IN_CLASSLIST
+/** GEMPLP (Generalized Max-product LP Relaxation) inference for fatcor graph
+ *
+ * Please refer to following paper for more details:
+ *
+ * [1] Fixing max-product: Convergent message passing algorithms for MAP LP-relaxations
+ * Amir Globerson, Tommi Jaakkola
+ * Advances in Neural Information Processing Systems (NIPS). Vancouver, Canada. 2007.
+ *
+ * [2] Approximate Inference in Graphical Models using LP Relaxations.
+ * David Sontag
+ * Ph.D. thesis, Massachusetts Institute of Technology, 2010.
+ *
+ * The original implementation of GEMPLP can be found:
+ * http://cs.nyu.edu/~dsontag/code/mplp_ver2.tgz
+ * http://cs.nyu.edu/~dsontag/code/mplp_ver1.tgz
+ */
+IGNORE_IN_CLASSLIST class CGEMPLP: public CMAPInferImpl
+{
+public:
+	/** Parameter for GEMPLP */
+	struct Parameter
+	{
+		Parameter(const int32_t max_iter = 1000,
+		          const float64_t obj_del_thr = 0.0002,
+		          const float64_t int_gap_thr = 0.0002)
+			: m_max_iter(max_iter),
+			  m_obj_del_thr(obj_del_thr),
+			  m_int_gap_thr(int_gap_thr)
+		{}
+
+		/** maximum number of outer iterations*/
+		int32_t m_max_iter;
+		/** threshold of the delta objective value */
+		float64_t m_obj_del_thr;
+		/** threshold of objective gap betweeb current and best integer assignment */
+		float64_t m_int_gap_thr;
+	};
+
+public:
+	/** Constructor */
+	CGEMPLP();
+
+	/** Constructor
+	 *
+	 * @param fg factor graph
+	 * @param param parameters
+	 */
+	CGEMPLP(CFactorGraph* fg, Parameter param = Parameter());
+
+	/** Destructor */
+	virtual ~CGEMPLP();
+
+	/** @return class name */
+	virtual const char* get_name() const
+	{
+		return "GEMPLP";
+	}
+
+	/** Inference
+	 *
+	 * @param assignment the assignment
+	 * @return the total energy after doing inference
+	 */
+	virtual float64_t inference(SGVector<int32_t> assignment);
+
+private:
+	/** Initialize GEMPLP with factor graph */
+	void init();
+	
+	/** Message updating on a clique
+	 *
+	 * Please refer to "GEMPLP" in NIPS paper of
+	 * A. Globerson and T. Jaakkola [1] for more details.
+	 *	
+	 */
+	void update_messages(int32_t id_clique);	
+
+public:
+	/** Computer the maximum value along the sub-dimension
+	 *
+	 * @param tar_arr target nd array
+	 * @param subset_inds sub-dimension indices
+	 * @param max_res the result nd array
+	 */
+	void max_in_subdimension(SGNDArray<float64_t> tar_arr, SGVector<int32_t> &subset_inds, SGNDArray<float64_t> &max_res) const;
+
+	/** Find separators between clqiues
+	 *
+	 * @param clique_A clique A
+	 * @param clique_B clique B
+	 * @return index in all separators
+	 */
+	int32_t find_separator_index(SGVector<int32_t> clique_A, SGVector<int32_t> clique_B);
```

Comment:
I guess this is analogy to sepset for clique trees. However, actually messages are passed between clusters/regions which are not formed in a hypertree but in general a hypergraph. I don't know if seperator set is a conventional in the context or not. Besides, calling it intersection sets is not a bad idea. The more we follow the OI names, the easier for people to understand.